CRAMv3.tex

```diff
@@ -1319,440 +1529,347 @@ \subsubsection*{BAM index}
-\indent $L \leq x < bL-1$
-}
+If we have integer values in the range 10 to 15 inclusive, the largest value would traditionally need 4 bits, but with an offset of -10 we can hold values 0 to 5, using a fixed size of 3 bits.
+Using fixed Offset and Length coming from the beta parameters, we decode these values as:
 
+\begin{tabular}{|>{\raggedright}p{105pt}|>{\raggedright}p{105pt}|>{\raggedright}p{105pt}|>{\raggedright}p{105pt}|}
+\hline
+Offset & Length & \textbf{Bits} & \textbf{Value}\tabularnewline
+\hline
+-10 & 3 & 000 & 10\tabularnewline
+\hline
+-10 & 3 & 001 & 11\tabularnewline
+\hline
+-10 & 3 & 010 & 12\tabularnewline
+\hline
+-10 & 3 & 011 & 13\tabularnewline
+\hline
+-10 & 3 & 100 & 14\tabularnewline
+\hline
+-10 & 3 & 101 & 15\tabularnewline
+\hline
+\end{tabular}
 
-where $b$ is the base and $L$ is the lower-bound.
+\subsection{Subexponential coding: codec ID 7}
 
-We ensure this property is true before every use of $C$ and after every
-use of $D$.  Finally to end the stream we flush any remaining data out
-by storing the end state of $x$.
+Can encode types \textit{Integer}.
 
+\subsubsection*{Definition}
 
-\textbf{Implementation specifics}
+Subexponential coding\footnote{Fast progressive lossless image compression, Paul G. Howard and Jeffrey Scott Vitter, 1994. \url{http://www.ittc.ku.edu/~jsv/Papers/HoV94.progressive_FELICS.pdf}} is parametrized by a non-negative integer $k$.
+For values $n < 2^{k+1}$ subexponential coding produces codewords identical to Rice coding \footnote{\url{https://en.wikipedia.org/wiki/Golomb_coding\#Rice_coding}}.  For larger values it grows logarithmically with $n$.
 
-We use an unsigned 32-bit integer to hold $x$. In encoding it is
-initialised to zero. For decoding it is read little-endian from the
-input stream.
+\subsubsection*{Encoding}
 
-Recall $freq_{i}$ is the frequency of the $i$-th symbol $s_{i}$ in alphabet
-$\mathbb{A}$.  We define $cfreq_i$ to be cumulative frequency of all symbols
-up to but not including $s_{i}$:
+\begin{enumerate}
+\item Add $\mathit{offset}$ to $n$.
+
+\item Determine $u$ and $b$ values from $n$
+\begin{align*}
+b =
+\begin{cases}
+  \ k                        & \text{ if $n < 2^k$} \\
+  \ \lfloor log_{2}n \rfloor & \text{ if $n \ge 2^k$}
+\end{cases}
+&\
+&u =
+\begin{cases}
+  \ 0     & \text{ if $n < 2^k$} \\
+  \ b-k+1 & \text{ if $n \ge 2^k$}
+\end{cases}
+\end{align*}
+
+\item Write $u$ in unary form; $u$ 1 bits followed by a single 0 bit.
+
+\item Write the bottom $b$-bits of $n$ in binary form.
+\end{enumerate}
 
-{
-\setlength{\parindent}{1cm}
-$ cfreq_{i} = \left\{
-\begin{array}{l l}
-0 & \quad \textrm{if $i < 1$} \\
-cfreq_{i-1} + freq_{i-1} & \quad \textrm{if $i \geq 1$}
-\end{array}
-\right. $
-% \\*[8pt]
-% \indent $cfreq_{i} = \displaystyle \sum_{j=0}^{i-1} freq_j$
-}
+\subsubsection*{Decoding}
 
+\begin{enumerate}
+\item Read $i$ in unary form, counting the number of leading 1s (prefix) in the codeword (discard the trailing 0 bit).
```

Comment:
Can we make this `u` as well, since it's the same as above?

Reply:
Done.